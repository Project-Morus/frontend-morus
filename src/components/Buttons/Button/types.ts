import { ButtonHTMLAttributes } from "react";

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  text: string;
  maxWidth?: number;
  variant?: VariantsTypes;
  isLoading?: boolean;
  disabled?: boolean;
}

export type VariantsTypes = "primary" | "secondary" | "red" | "green";
