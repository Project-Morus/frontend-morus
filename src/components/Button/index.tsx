import { ReactNode } from "react";
import { IButtonProps, VariantsTypes } from "./types";
import ButtonPrimary from "./variants/Primary";
import ButtonSecondary from "./variants/Secondary";
import ButtonRed from "./variants/Red";
import ButtonGreen from "./variants/Green";

const Button = ({ variant = "primary", ...props }: IButtonProps) => {
  const variantType: Record<VariantsTypes, ReactNode> = {
    primary: <ButtonPrimary {...props} />,
    secondary: <ButtonSecondary {...props} />,
    red: <ButtonRed {...props} />,
    green: <ButtonGreen {...props} />,
  };

  return variantType[variant];
};

export default Button;
