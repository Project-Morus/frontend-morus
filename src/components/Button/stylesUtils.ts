import { styled } from "styled-components";

export const BASE_STYLE = styled.button<{ $maxWidth?: number }>`
  width: 100%;

  font-weight:  ${({ theme }) => theme.typography.FONT_WEIGHTS.bold};

  border: none;
  border-radius: ${({ theme }) => theme.border.radius.regular};

  background: none;

  cursor: pointer;

  padding: ${({ theme }) => theme.spacing[4]};

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }

  transition: all 0.2s;
  
  ${({ $maxWidth }) => $maxWidth && `max-width: ${$maxWidth}px`};
`
