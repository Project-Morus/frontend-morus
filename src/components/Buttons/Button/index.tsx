import { ReactNode } from "react";
import { IButtonProps, VariantsTypes } from "./types";
import { ButtonGreen, ButtonPrimary, ButtonRed, ButtonSecondary } from "./variants";

const Button = ({ variant = "primary", ...props }: IButtonProps) => {
  const variantType: Record<VariantsTypes, ReactNode> = {
    primary: <ButtonPrimary {...props} />,
    secondary: <ButtonSecondary {...props} />,
    red: <ButtonRed {...props} />,
    green: <ButtonGreen {...props} />,
  };

  return variantType[variant];
};

export default Button;
