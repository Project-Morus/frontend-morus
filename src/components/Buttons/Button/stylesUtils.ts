import { styled } from "styled-components";

export const HEIGHT_BUTTON = '50px'

export const BASE_STYLE = styled.button<{ $maxWidth?: number }>`
  width: 100%;
  height: ${HEIGHT_BUTTON};

  display: flex;
  align-items: center;
  justify-content: center;

  font-weight:  ${({ theme }) => theme.typography.FONT_WEIGHTS.bold};

  border: none;
  border-radius: ${({ theme }) => theme.border.radius.regular};

  background: none;

  cursor: pointer;

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }

  transition: all 0.2s;
  
  ${({ $maxWidth }) => $maxWidth && `max-width: ${$maxWidth}px`};
`
