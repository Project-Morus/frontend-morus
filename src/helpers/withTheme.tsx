import { ComponentType, JSX } from 'react';
import { ThemeProvider } from "styled-components";
import { defaultTheme } from "../styles/themes/default";

type ComponentPropsType<T> = T & JSX.IntrinsicAttributes

export function withTheme<T>(WrapperComponent: ComponentType<ComponentPropsType<T>>) {
  return (props: ComponentPropsType<T>) => {
    return (
      <ThemeProvider theme={defaultTheme}>
        <WrapperComponent {...props} />
      </ThemeProvider>
    );
  };
}

