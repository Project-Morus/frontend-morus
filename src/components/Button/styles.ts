import { styled } from "styled-components";
import { BASE_STYLE } from "./stylesUtils";


export const ButtonPrimarySC = styled(BASE_STYLE)`
  background-color: ${({ theme }) => theme.colors.primary[500]};
  color: ${({ theme }) => theme.colors.white[50]};

  &:not(:disabled):hover {
    background-color: ${({ theme }) => theme.colors.primary[600]};
  }
`

export const ButtonSecondarySC = styled(BASE_STYLE)`
  border: ${({ theme }) => theme.border.width.bold} solid ${({ theme }) => theme.colors.primary[500]};
  color: ${({ theme }) => theme.colors.primary[500]};

  &:not(:disabled):hover {
    border-color: ${({ theme }) => theme.colors.primary[600]};
    background-color: ${({ theme }) => theme.colors.primary[500]};
    color: ${({ theme }) => theme.colors.white[50]};
  }

  &:disabled {
    opacity: 0.5; 
    cursor: not-allowed;
  }
`