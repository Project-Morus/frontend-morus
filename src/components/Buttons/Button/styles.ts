import { styled } from "styled-components";
import { BASE_STYLE } from "./stylesUtils";

export const ButtonPrimarySC = styled(BASE_STYLE)`
  background-color: ${({ theme }) => theme.colors.primary[500]};
  color: ${({ theme }) => theme.colors.white[50]};

  &:not(:disabled):hover {
    background-color: ${({ theme }) => theme.colors.primary[600]};
  }
`;

export const ButtonSecondarySC = styled(BASE_STYLE)`
  border: ${({ theme }) => theme.border.width.bold} solid ${({ theme }) => theme.colors.primary[500]};
  color: ${({ theme }) => theme.colors.primary[500]};

  &:not(:disabled):hover {
    border-color: ${({ theme }) => theme.colors.primary[600]};
    background-color: ${({ theme }) => theme.colors.primary[500]};
    color: ${({ theme }) => theme.colors.white[50]};
  }
`;

export const ButtonRedSC = styled(BASE_STYLE)`
  border: ${({ theme }) => theme.border.width.bold} solid ${({ theme }) => theme.colors.red[500]};
  background-color: ${({ theme }) => theme.colors.red[500]};
  color: ${({ theme }) => theme.colors.white[50]};

  &:not(:disabled):hover {
    border-color: ${({ theme }) => theme.colors.red[600]};
    background-color: ${({ theme }) => theme.colors.red[600]};
    color: ${({ theme }) => theme.colors.white[50]};
  }
`;

export const ButtonGreenSC = styled(BASE_STYLE)`
  border: ${({ theme }) => theme.border.width.bold} solid ${({ theme }) => theme.colors.green[500]};
  background-color: ${({ theme }) => theme.colors.green[500]};
  color: ${({ theme }) => theme.colors.white[50]};

  &:not(:disabled):hover {
    border-color: ${({ theme }) => theme.colors.green[600]};
    background-color: ${({ theme }) => theme.colors.green[600]};
    color: ${({ theme }) => theme.colors.white[50]};
  }
`;