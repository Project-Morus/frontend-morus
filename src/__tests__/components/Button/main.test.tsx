import { render, screen, fireEvent } from '@testing-library/react'
import Button from '../../../components/Button'
import { withTheme } from '../../../helpers/withTheme'

const ButtonWithTheme = withTheme(Button)
const buttonElement = 'button'
const textButton = 'Button'
describe('Button Component', () => {
  test('Verify if component is rendering', () => {
    render(<ButtonWithTheme text={textButton} />)


    expect(screen.getByRole(buttonElement)).toBeTruthy()
  })

  test('Verify if text is rendering in Button', () => {
    render(<ButtonWithTheme text={textButton} />)

    expect(screen.getByText(textButton))
  })

  test('Receives a "click" and calls an onClick function', () => {
    const onClickTest = jest.fn()
    render(<ButtonWithTheme text={textButton} onClick={onClickTest} />)

    fireEvent.click(screen.getByRole(buttonElement))
    expect(onClickTest).toHaveBeenCalledTimes(1)
  })

  test('OnClick is not fired when the button is disabled', () => {
    const onClickTest = jest.fn()

    render(<ButtonWithTheme text={textButton} onClick={onClickTest} disabled />)

    fireEvent.click(screen.getByRole('button'))
    expect(onClickTest).not.toHaveBeenCalledTimes(1)
  })


  test("Verify that the button is disabled", () => {
    render(<ButtonWithTheme text={textButton} disabled />);

    expect(screen.getByRole(buttonElement)).toBeDisabled();
  });

  test('Snapshot Button', () => {
    const { asFragment } = render(<ButtonWithTheme text={textButton} />)

    expect(asFragment()).toMatchSnapshot()
  })
  test('Snapshot Button with variant secondary', () => {
    const { asFragment } = render(<ButtonWithTheme text={textButton} variant='secondary' />)

    expect(asFragment()).toMatchSnapshot()
  })
})