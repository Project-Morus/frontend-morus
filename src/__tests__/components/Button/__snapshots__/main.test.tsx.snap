// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Snapshot Button 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  font-weight: 700;
  border: none;
  border-radius: 5px;
  background: none;
  cursor: pointer;
  padding: 0.75rem;
  transition: all 0.2s;
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c1 {
  background-color: #0353A4;
  color: #FFF;
}

.c1:not(:disabled):hover {
  background-color: #034C95;
}

<button
    class="c0 c1"
  >
    <span>
      Button
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button Component Snapshot Button with variant secondary 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  font-weight: 700;
  border: none;
  border-radius: 5px;
  background: none;
  cursor: pointer;
  padding: 0.75rem;
  transition: all 0.2s;
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c1 {
  border: 2px solid #0353A4;
  color: #0353A4;
}

.c1:not(:disabled):hover {
  border-color: #034C95;
  background-color: #0353A4;
  color: #FFF;
}

.c1:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

<button
    class="c0 c1"
  >
    <span>
      Button
    </span>
  </button>
</DocumentFragment>
`;
